ユーザー管理機能

iejima0429 opened the pull request:
#What
ユーザー管理機能を作成

#Why
ユーザーの新規登録、ログアウト、ログインを実装するため

↓新規登録成功動画
https://gyazo.com/fb247d127096bc71c7705e72af210194

↓ログアウト成功、ログイン失敗・成功動画
https://gyazo.com/ed7dab1c8296cd047ff2e0256eb48993

↓新規登録失敗動画
https://gyazo.com/adf391ed0da5943a60a72b18f90c51f7
https://gyazo.com/a667c57b93b895482b258e15c16d40a0

↓テスト成功画像
https://gyazo.com/5f66e3b10dbeeccf75cb049f135b317c

iejima0429 committed:
新規登録ボタンを押すと遷移する

iejima0429 committed:
新規登録できるようにした

iejima0429 committed:
ログインできるようにした

iejima0429 committed:
カナ入力、パスワード英数混合実装

iejima0429 committed:
エラーメッセージ表示できるように修正

iejima0429 committed:
モデルの単体テストを記述

AsukaToyoshima commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+         :recoverable, :rememberable, :validatable
+
+  validates :nickname, presence: true
+  validates :first_name, presence: true
+  validates :last_name, presence: true
```
今回の要件を実現するため、下記のバリデーションを追加しましょう。
```
- ユーザー本名は、全角（漢字・ひらがな・カタカナ）での入力が必須であること
```

AsukaToyoshima commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+  validates :nickname, presence: true
+  validates :first_name, presence: true
+  validates :last_name, presence: true
+  validates :first_name_kana, presence: true, format: { with: /\A[\p{katakana}\p{blank}ー－]+\z/ }
+  validates :last_name_kana, presence: true, format: { with: /\A[\p{katakana}\p{blank}ー－]+\z/ }
```
こちらの正規表現を以下のように修正しましょう。
```Ruby
/\A[\p{katakana}ー]+\z/
```
設定いただいている正規表現では、空白スペースも許可してしまいます。

AsukaToyoshima commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,96 @@
+      @user.password_confirmation = @user.password
+      @user.valid?
+      expect(@user.errors.full_messages).to include("Password is invalid")
+    end
```
今回の要件、設定したバリデーションに合わせ、以下のテストを追加しましょう。
```
・メールアドレスに@を含まない場合は登録できない
・英字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない
・姓（全角）に半角文字が含まれていると登録できない
・名（全角）に半角文字が含まれていると登録できない
```

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
初回から完成度の高いコードが作成できておりますね！
3点コメントさせていただきました。ご確認の上、修正をお願いいたします。

疑問点・不明点等ございましたら、コネクトメンターへお気軽にご質問ください。

iejima0429 committed:
モデルの単体テスト修正

・メールアドレスに@を含まない場合は登録できない
・英字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない
・姓（全角）に半角文字が含まれていると登録できない
・名（全角）に半角文字が含まれていると登録できない

主に上記を修正しました。

iejima0429 commented:
↓テスト成功画像
https://gyazo.com/51a41b6865affd23ae21425366c81ed8

ShotaOtake commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
+  devise_for :users
   root to: "items#index"
+  resources :users
```
◎下記については削除しておきましょう！
```
resources :users
```

devise関連のルーティングはdevise_for :usersにて全て用意されているためです。

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

完成まであと一歩です！
引き続き実装頑張っていきましょう！

iejima0429 committed:
route.rbのresources :users削除

無駄な記述のため削除しました。

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます！実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう！
